File: 프로그래머스/3/43164. 여행경로/여행경로.py
import sys
from collections import defaultdict
sys.setrecursionlimit(10**6)
visited = set()
def solution(tickets):
    
    tickets.sort(key = lambda x:(x[0],x[1]))
    # print(tickets)
        
    ans = ["ICN"]

    answer = []
    def dfs(x):
        if len(visited)==len(tickets):
            answer.append(ans[:])
            return
        
        for i in range(len(tickets)):
            if tickets[i][0] == x and i not in visited:
                visited.add(i)
                ans.append(tickets[i][1])
                dfs(tickets[i][1])
                ans.pop()
                visited.remove(i)
        # return
    dfs("ICN")
    
    return answer[0]

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;
class Solution {
    static List<String> results;
    public String[] solution(String[][] tickets) {
        results = new ArrayList<>();
        //tickets 정렬 하기


        dfs("ICN",tickets,new boolean[tickets.length],0,"ICN");
        
        Collections.sort(results);
        String[] answer = results.get(0).split(" ");

        return answer;
    }
    
    public void dfs(String start, String[][] tickets, boolean[] visited, int cnt,String tmp){
        if(cnt == tickets.length){
            results.add(tmp);
        }
        else{
            //현재 출발지에서 출발하는 티켓이 있는지 확인
            for(int idx=0;idx<tickets.length;idx++){
                if(tickets[idx][0].equals(start) && !visited[idx]){
                    visited[idx] = true;
                    dfs(tickets[idx][1],tickets,visited,cnt+1,tmp +" "+tickets[idx][1]);
                    visited[idx] = false;
                }
            }
        }
    }
}